File: web/src/lib/geometry/spiral.ts
import type { Polygon, LayerMap, Port, GeometryResult, SpiralInductorParams } from './types';
import { viaGrid, pgs4 } from './utils';
import { makeAspectShiftY, mapY } from './primitives';

/** Build spiral inductor geometry. Polygons are the single source of truth. */
export function buildSpiralInductor(params: SpiralInductorParams): GeometryResult {
	const { Dout, N, sides, width, spacing, via_spacing, via_width, via_in_metal } = params;
	const ar = params.aspectRatio ?? 1;

	const PI = Math.PI;
	const v = width / Math.cos(PI / sides);
	const s = (spacing + width) / Math.cos(PI / sides);

	let R1 = Dout / 2 / Math.cos(PI / sides);
	let R2 = R1 - v;

	const nPts = sides / 2;
	const angles: number[] = [];
	for (let i = 0; i < nPts; i++) {
		angles.push(PI * (1 / (2 * nPts) + i * (1 - 1 / nPts) / (nPts - 1)));
	}

	const extend = 2 * (via_width + via_in_metal) + via_spacing;
	const xShift = -s / 2 * Math.cos(PI / sides);
	const yShift = -s / 2 * Math.sin(PI / sides);

	const opposite = params.portSide === 'opposite';
	// For opposite-side ports: spare one half-turn so the spiral ends on the right side
	// at a larger inner radius, leaving room for the bridge to cross left
	const nSections = opposite ? 2 * N - 1 : 2 * N;

	// --- Compute outer/inner traces ---
	const xOut: number[] = [];
	const yOut: number[] = [];
	const xIn: number[] = [];
	const yIn: number[] = [];

	let r1 = R1, r2 = R2;
	for (let section = 0; section < nSections; section++) {
		if (section % 2 === 0) {
			for (const phi of angles) {
				xOut.push(r1 * Math.cos(phi));
				xIn.push(r2 * Math.cos(phi));
				yOut.push(r1 * Math.sin(phi));
				yIn.push(r2 * Math.sin(phi));
			}
		} else {
			for (const phi of angles) {
				xOut.push(-r1 * Math.cos(phi) + xShift);
				xIn.push(-r2 * Math.cos(phi) + xShift);
				yOut.push(-r1 * Math.sin(phi) + yShift);
				yIn.push(-r2 * Math.sin(phi) + yShift);
			}
		}
		r1 -= s / 2;
		r2 -= s / 2;
	}

	// For opposite-side: both ports centered at y=0 (no offset needed since ports are on different sides)
	// For same-side: entry above, exit below (offset by spacing to avoid shorting)
	const entryYCenter = opposite ? 0 : (width + spacing) / 2;
	const exitYCenter = opposite ? 0 : -(width + spacing) / 2;

	// Start connector (P1 always enters from right)
	const xOutStart = [Dout / 2 + width, xOut[0]];
	const xInStart = [Dout / 2 + width, xIn[0]];
	const yOutStart = [entryYCenter + width / 2, entryYCenter + width / 2];
	const yInStart = [entryYCenter - width / 2, entryYCenter - width / 2];

	// End connector
	const xOutEnd = [xOut[xOut.length - 1]];
	const xInEnd = [xIn[xIn.length - 1]];
	const yOutEnd = [opposite ? -width / 2 : -spacing / 2];
	const yInEnd = [opposite ? -width / 2 : -spacing / 2];

	// Full winding polygon
	const xPoly = [
		...xOutStart, ...xOut, ...xOutEnd,
		...[...xInEnd].reverse(), ...[...xIn].reverse(), ...[...xInStart].reverse()
	];
	const yPoly = [
		...yOutStart, ...yOut, ...yOutEnd,
		...[...yInEnd].reverse(), ...[...yIn].reverse(), ...[...yInStart].reverse()
	];
	const windingPolygon: Polygon = { x: xPoly, y: yPoly };

	// Underpass polygon
	const lastXIn = xIn[xIn.length - 1];
	const lastXOut = xOut[xOut.length - 1];
	const underpassEndX = opposite ? -(Dout / 2 + width) : Dout / 2 + width;
	const underpassPolygon: Polygon = {
		x: [lastXIn, underpassEndX, underpassEndX, lastXIn],
		y: [exitYCenter - width / 2, exitYCenter - width / 2, exitYCenter + width / 2, exitYCenter + width / 2],
	};

	// Via polygons
	const viaCenterX = lastXOut + (lastXIn - lastXOut) / 2;
	const viaCenterY = exitYCenter;

	let viaPolys: Polygon[];
	if (extend > width) {
		viaPolys = viaGrid(
			viaCenterX, viaCenterY + (extend - width) / 2,
			width - 2 * via_in_metal, extend - 2 * via_in_metal,
			via_spacing, via_width
		);
	} else {
		viaPolys = viaGrid(
			viaCenterX, viaCenterY,
			width - 2 * via_in_metal, width - 2 * via_in_metal,
			via_spacing, via_width
		);
	}

	// Apply aspect ratio — stretch the straight sides only (y=0 fixed).
	const shiftY = makeAspectShiftY(Dout, ar);
	const layers: LayerMap = {
		windings: [mapY(windingPolygon, shiftY)],
		crossings: [mapY(underpassPolygon, shiftY)],
		vias: viaPolys.map(p => mapY(p, shiftY)),
		pgs: [],
	};

	const ports: Port[] = [
		{ name: 'P1', x: Dout / 2 + width, y: shiftY(entryYCenter), layer: 'windings', role: 'signal' },
		{ name: 'P2', x: underpassEndX, y: shiftY(exitYCenter), layer: 'crossings', role: 'signal' },
	];

	return { layers, ports };
}

export function isSpiralValid(params: SpiralInductorParams): boolean {
	const { Dout, N, sides, width, spacing, via_spacing, via_width, via_in_metal } = params;
	const opposite = params.portSide === 'opposite';
	const extend = 2 * (via_width + via_in_metal) + via_spacing;
	if (extend > width) return false;
	if (opposite && N < 2) return false;

	if (opposite) {
		// Opposite-side: 2*N - 1 sections, underpass centered at y=0
		const s = (spacing + width) / Math.cos(Math.PI / sides);
		const R1 = Dout / 2 / Math.cos(Math.PI / sides);
		const innerR = R1 - (2 * N - 1) * s / 2;
		if (innerR * Math.sin(Math.PI / sides) < width / 2) return false;
	} else {
		// Same-side: original validation
		const Din = Dout - (N + 1) * (width + spacing);
		if (Math.abs(Din / 2 * Math.atan(Math.PI / sides)) < width + spacing / 2) return false;
	}

	return true;
}

export function addPgs(layers: LayerMap, D: number, w: number, s: number): void {
	layers.pgs = pgs4(D, w, s);
}

File: rapidpassives/geometry/spiral.py
"""Spiral inductor geometry — Python port of web/src/lib/geometry/spiral.ts.

Polygons are the single source of truth, exactly as in the TS app, so the FEM
JSON produced here matches the web export for identical parameters.
"""
from __future__ import annotations

import math

from .primitives import Poly, make_aspect_shift_y, map_y, pgs4, via_grid
from .result import GeometryResult, Port


def build_spiral_inductor(params: dict) -> GeometryResult:
    Dout = params["Dout"]; N = params["N"]; sides = params["sides"]
    width = params["width"]; spacing = params["spacing"]
    via_spacing = params["via_spacing"]; via_width = params["via_width"]
    via_in_metal = params["via_in_metal"]
    ar = params.get("aspectRatio", 1.0)
    opposite = params.get("portSide") == "opposite"

    PI = math.pi
    s = (spacing + width) / math.cos(PI / sides)
    v = width / math.cos(PI / sides)
    R1 = Dout / 2 / math.cos(PI / sides)
    R2 = R1 - v

    n_pts = sides // 2
    angles = [PI * (1 / (2 * n_pts) + i * (1 - 1 / n_pts) / (n_pts - 1)) for i in range(n_pts)]

    extend = 2 * (via_width + via_in_metal) + via_spacing
    x_shift = -s / 2 * math.cos(PI / sides)
    y_shift = -s / 2 * math.sin(PI / sides)

    n_sections = 2 * N - 1 if opposite else 2 * N

    x_out: list[float] = []; y_out: list[float] = []
    x_in: list[float] = []; y_in: list[float] = []
    r1, r2 = R1, R2
    for section in range(n_sections):
        if section % 2 == 0:
            for phi in angles:
                x_out.append(r1 * math.cos(phi)); x_in.append(r2 * math.cos(phi))
                y_out.append(r1 * math.sin(phi)); y_in.append(r2 * math.sin(phi))
        else:
            for phi in angles:
                x_out.append(-r1 * math.cos(phi) + x_shift); x_in.append(-r2 * math.cos(phi) + x_shift)
                y_out.append(-r1 * math.sin(phi) + y_shift); y_in.append(-r2 * math.sin(phi) + y_shift)
        r1 -= s / 2; r2 -= s / 2

    entry_yc = 0.0 if opposite else (width + spacing) / 2
    exit_yc = 0.0 if opposite else -(width + spacing) / 2

    x_out_start = [Dout / 2 + width, x_out[0]]
    x_in_start = [Dout / 2 + width, x_in[0]]
    y_out_start = [entry_yc + width / 2, entry_yc + width / 2]
    y_in_start = [entry_yc - width / 2, entry_yc - width / 2]

    x_out_end = [x_out[-1]]
    x_in_end = [x_in[-1]]
    y_end = [-width / 2 if opposite else -spacing / 2]

    x_poly = x_out_start + x_out + x_out_end + list(reversed(x_in_end)) + list(reversed(x_in)) + list(reversed(x_in_start))
    y_poly = y_out_start + y_out + y_end + list(reversed(y_end)) + list(reversed(y_in)) + list(reversed(y_in_start))
    winding_polygon: Poly = list(zip(x_poly, y_poly))

    last_x_in = x_in[-1]; last_x_out = x_out[-1]
    underpass_end_x = -(Dout / 2 + width) if opposite else Dout / 2 + width
    underpass_polygon: Poly = [
        (last_x_in, exit_yc - width / 2),
        (underpass_end_x, exit_yc - width / 2),
        (underpass_end_x, exit_yc + width / 2),
        (last_x_in, exit_yc + width / 2),
    ]

    via_cx = last_x_out + (last_x_in - last_x_out) / 2
    via_cy = exit_yc
    if extend > width:
        via_polys = via_grid(via_cx, via_cy + (extend - width) / 2,
                             width - 2 * via_in_metal, extend - 2 * via_in_metal, via_spacing, via_width)
    else:
        via_polys = via_grid(via_cx, via_cy,
                             width - 2 * via_in_metal, width - 2 * via_in_metal, via_spacing, via_width)

    shift_y = make_aspect_shift_y(Dout, ar)
    layers = {
        "windings": [map_y(winding_polygon, shift_y)],
        "crossings": [map_y(underpass_polygon, shift_y)],
        "vias": [map_y(p, shift_y) for p in via_polys],
        "pgs": [],
    }
    ports = [
        Port("P1", Dout / 2 + width, shift_y(entry_yc), "windings"),
        Port("P2", underpass_end_x, shift_y(exit_yc), "crossings"),
    ]
    return GeometryResult(layers, ports)


def add_pgs(result: GeometryResult, D: float, w: float, s: float) -> None:
    result.layers["pgs"] = pgs4(D, w, s)
